rfd/0093-offline-access.md

```diff
@@ -0,0 +1,190 @@
+2) `types.SessionTracker`
+
+
+###### Per-Session MFA
+`tsh` determines whether per-session MFA is required **prior** to connecting to a node. Which means that every invocation of `tsh ssh` requires a connection to the Auth server. 
+
+Successful connection attempt when Auth is available:
+```mermaid
+sequenceDiagram
+    participant tsh
+    participant Auth
+    participant Node
+
+    tsh->>+Auth: IsMFARequired
+    Auth-->>-tsh: yes
+    tsh->>+Auth: perform mfa ceremony
+    Auth-->>+tsh: issue challenge
+    tsh-->>+Auth: challenge response
+    Auth-->>+tsh: issue certificates
+    tsh->>+Node: Connect
+    Node-->>+tsh: Success
+```
+
+Failed connection attempt when Auth is unavailable:
+```mermaid
+sequenceDiagram
+    participant tsh
+    participant Auth
+    participant Node
+
+    tsh->>+Auth: IsMFARequired
+    Auth-->>-tsh: Error
+```
+
+However, the decision if the connection should be established is ultimately made by the node, not the Auth server. The node
+performs an RBAC check on all inbound connections to determines if among other things per-session MFA is required and the MFA ceremony was 
+performed. Thus, `tsh` should **always** attempt to connect to a node even if the `proto.AuthService/IsMFARequired` RPC fails. In the worst
+case this makes `tsh ssh` for per-session MFA enabled connections take longer to fail. But that cost comes at the ability to allow all
+users that will never have to perform per-session MFA to be able to connect to nodes without needing the Auth server if they already 
+have certificates.
+
+Successful connection attempt when Auth isn't available and per-session MFA is **not** required
+```mermaid
+sequenceDiagram
+    participant tsh
+    participant Auth
+    participant Node
+
+    tsh->>+Auth: IsMFARequired
+    Auth-->>-tsh: Error
+    tsh->>+Node: Connect
+    Node-->>+tsh: Success
+```
+
+Failed connection attempt when Auth is unavailable and per-session MFA is required:
+```mermaid
+sequenceDiagram
+    participant tsh
+    participant Auth
+    participant Node
+
+    tsh->>+Auth: IsMFARequired
+    Auth-->>-tsh: Error
+    tsh->>+Node: Connect
+    Node-->>+tsh: Access Denied
+```
+
+###### Session Trackers
+A new session tracker is created for every session. In the event that persisting the session tracker fails the 
+session is aborted. 
+
+Instead of immediately aborting the session if the session tracker cannot be persisted, the node could first check if 
+moderated sessions are required. In the event that they aren't the session should be allowed to proceed using a local 
+session tracker resource. 
+
+While this does allow access to the node, it will prevent the following:
+- the session from appearing in the active sessions list
+- the session will not be joinable by other users
+- the session recording will not be available
+
+
+## Security
+
+The changes to `tsh` should not impact security at all. They will still enforce that per-session MFA is performed if
```

Comment:
How would this feature interact with locking? The auth server may be running, but simply inaccessible by the node or client, in which case locks would not propagate appropriately.